tutorials/dataframe/df107_SingleTopAnalysis.py

```diff
@@ -0,0 +1,245 @@
+            }
+        }
+        if (type[i] == 13 && abs(eta[i]) < 2.5 && abs(trackd0pv[i] / tracksigd0pv[i]) < 3) {
+            const ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+            if (abs(z0[i] * sin(p.Theta())) < 0.5) {
+                if (idx == -1) idx = i;
+                else return -1;
+            }
+        }
+    }
+    return idx;
+}
+
+float ComputeTransverseWMass(float lep_pt, float lep_eta, float lep_phi, float lep_e, float met_et, float met_phi)
+{
+    const ROOT::Math::PtEtaPhiEVector lep(lep_pt, lep_eta, lep_phi, lep_e);
+    const ROOT::Math::PtEtaPhiEVector met(met_et, 0, met_phi, met_et);
+    return sqrt(2 * lep.Pt() * met.Et() * (1 - cos(ROOT::Math::VectorUtil::DeltaPhi(lep, met))));
+}
+""")
+
+
+for s in samples:
+    # Select events with electron or muon trigger and with a missing tranverse energy above 30 GeV
+    df[s] = df[s].Filter("trigE || trigM")\
+                 .Filter("met_et > 30000")
+
+    # Perform preselection of highly isolated leptons
+    df[s] = df[s].Define("goodlep", "lep_isTightID && lep_pt > 35000 && lep_ptcone30 / lep_pt < 0.1 && lep_etcone20 / lep_pt < 0.1")\
+                 .Filter("Sum(goodlep) > 0")
+
+    # Find a single good lepton, otherwise return -1 as index
+    df[s] = df[s].Define("idx_lep", "FindGoodLepton(goodlep, lep_type, lep_pt, lep_eta, lep_phi, lep_E, lep_trackd0pvunbiased, lep_tracksigd0pvunbiased, lep_z0)")\
+                 .Filter("idx_lep != -1")
+
+    # Compute transverse mass of the W boson using the missing transverse energy and the good lepton
+    # Use only events with a transverse mass of the reconstruced W boson larger than 60 GeV
+    df[s] = df[s].Define("mtw", "ComputeTransverseWMass(lep_pt[idx_lep], lep_eta[idx_lep], lep_phi[idx_lep], lep_E[idx_lep], met_et, met_phi)")\
+                 .Filter("mtw > 60000")
+
+    # Perform preselection of jets
+    df[s] = df[s].Filter("Sum(jet_pt > 30000 && abs(jet_eta) < 2.5) > 0")
+
+    # Select events with two good jets and one b-jet and find the indices in the collections
+    df[s] = df[s].Define("goodjet", "jet_pt > 60000 || abs(jet_eta) > 2.4 || jet_jvt > 0.59")\
+                 .Filter("Sum(goodjet) == 2")\
+                 .Define("goodbjet", "goodjet && jet_MV2c10 > 0.8244273")\
+                 .Filter("Sum(goodbjet) == 1")\
+                 .Define("idx_tagged", "ArgMax(goodjet && goodbjet)")\
+                 .Define("idx_untagged", "ArgMax(goodjet && !goodbjet)")
+
+    # Select events based on the jet kinematics and the scalar sum of the transverse momentum
+    # from the lepton, jets and met above 195 GeV
+    df[s] = df[s].Filter("abs(jet_eta[idx_untagged]) > 1.5 && abs(jet_eta[idx_tagged] - jet_eta[idx_untagged]) > 1.5")\
+                 .Filter("lep_pt[idx_lep] + jet_pt[idx_tagged] + jet_pt[idx_untagged] + met_et > 195000")
+
+# Compute luminosity, scale factors and MC weights for simulated events
+lumi = 10064.0
```

Comment:
Any way to get this from the data files? A real analysis would hopefully not hardcode this.

Reply:
Computing the lumi itself is not so trivial, no? At least not in CMS :D Well, anyway, we have to stick to hard coding here since the provided files don't have this info.